pkg/policy/distillery.go

```diff
@@ -117,10 +117,22 @@ func (cache *PolicyCache) updateSelectorPolicy(identity *identityPkg.Identity) (
 		return false, fmt.Errorf("SelectorPolicy not found in cache for ID %d", identity.ID)
 	}
 
-	// Don't resolve policy if it was already done for this Identity.
+	// As long as UpdatePolicy() is triggered from endpoint
+	// regeneration, it's possible for two endpoints with the
+	// *same* identity to race to update the policy here. Such
+	// racing would lead to first of the endpoints using a
+	// selectorPolicy that is already detached from the selector
+	// cache, and thus not getting any incremental updates.
+	//
+	// Lock the 'cip' for the duration of the revision check and
+	// the possible policy update.
+	cip.Lock()
+	defer cip.Unlock()
+
 	currentPolicy := cip.getPolicy()
-	currentRevision := currentPolicy.Revision
-	if revision <= currentRevision {
+
+	// Don't resolve policy if it was already done for this or later revision.
+	if revision <= currentPolicy.Revision {
```

Comment:
Why don't we get the `revision` after taking the lock of `cip`?

Reply:
It is not necessary as the repo is read locked during this call, but I changed to this like you suggested here to make this clearer.